pkg/gui/controllers/helpers/fixup_helper.go

```diff
@@ -0,0 +1,175 @@
+package helpers
+
+import (
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/jesseduffield/generics/set"
+	"github.com/jesseduffield/lazygit/pkg/commands/models"
+	"github.com/jesseduffield/lazygit/pkg/gui/types"
+	"github.com/jesseduffield/lazygit/pkg/utils"
+	"github.com/samber/lo"
+)
+
+type FixupHelper struct {
+	c *HelperCommon
+}
+
+func NewFixupHelper(
+	c *HelperCommon,
+) *FixupHelper {
+	return &FixupHelper{
+		c: c,
+	}
+}
+
+type deletedLineInfo struct {
+	filename     string
+	startLineIdx int
+	numLines     int
+}
+
+func (self *FixupHelper) HandleFindBaseCommitForFixupPress() error {
+	diff, useIndex, err := self.getDiff()
+	if err != nil {
+		return err
+	}
+	if diff == "" {
+		return self.c.ErrorMsg("No changes to commit")
+	}
+
+	deletedLineInfos := self.parseDiff(diff)
+	if len(deletedLineInfos) == 0 {
+		return self.c.ErrorMsg("No deleted lines in diff")
+	}
+
+	shas := self.blameDeletedLines(deletedLineInfos)
+
+	if len(shas) == 0 {
+		// This should never happen
+		return self.c.ErrorMsg("No base commits found")
+	}
+	if len(shas) > 1 {
+		subjects, err := self.c.Git().Commit.GetShasAndCommitMessagesFirstLine(shas)
+		if err != nil {
+			return err
+		}
+		return self.c.ErrorMsg("Multiple base commits found. (" +
+			lo.Ternary(useIndex, "Try staging fewer changes", "Try staging some of the changes") +
+			")\n\n" + subjects)
+	}
+
+	commit, index, ok := lo.FindIndexOf(self.c.Model().Commits, func(commit *models.Commit) bool {
+		return commit.Sha == shas[0]
+	})
+	if !ok {
+		if self.c.Model().Commits[len(self.c.Model().Commits)-1].Status == models.StatusMerged {
+			// If the commit is not found, it's most likely because it's already
+			// merged, and more than 300 commits away. Check if the last known
+			// commit is already merged; if so, show the "already merged" error.
+			return self.c.ErrorMsg("The base commit for this change is already on master")
+		}
+		// If we get here, the current branch must have more then 300 commits. Unlikely...
+		return self.c.ErrorMsg("Base commit is not in current view")
+	}
+	if commit.Status == models.StatusMerged {
+		return self.c.ErrorMsg("The base commit for this change is already on master")
+	}
+
+	if !useIndex {
```

Comment:
Took me a bit to make sense of this part: would be good to add a comment saying something like 'If there are no existing staged changes, then we assume the user intends to amend the commit with all files, so we stage them before focusing the commits view'.

In fact I'd change `useIndex` to be named `noStagedChanges` to make this clearer.

Reply:
I dislike variables that express a negated state, so I went with `hasStagedChanges` instead; hope that still works for you. (Changed in 8b76a81e6bab)